### DISCUSSION TITLE
Type checking of pandas `series.apply`

### DISCUSSION THREAD

**Discussion by JanEricNitschke** [microsoft/pyright#5137] [General]
I am currently adding type checking to some of my code that uses pandas and have run into an issue where i am not sure if its a mistake on my side, incomplete/inaccurate type stubs or pyright (or something else).

I have a dataframe which has in one of its columns a series of dictionaries.

Now i am trying to transform all of these dictionaries into dataframes themselves.

So i am doing:

```
dataframe["position_df"] = dataframe["position_df"].apply(
            self.__transform_to_data_frame
        )
```

where basically

```
    def __transform_to_data_frame(
        self, json_format_dict: dict
    ) -> pd.DataFrame:
        return pd.DataFrame(json_format_dict)
```

This all works nice at run time but i am getting a pyright type checking error:

```
error: Argument of type "(json_format_dict: PositionDatasetJSON) -> DataFrame" cannot be assigned to parameter "func" of type "(...) -> Series[Unknown]" in function "apply"
    Type "(json_format_dict: PositionDatasetJSON) -> DataFrame" cannot be assigned to type "(...) -> Series[Unknown]"
      Function return type "DataFrame" is incompatible with type "Series[Unknown]"
        "DataFrame" is incompatible with "Series[Unknown]" (reportGeneralTypeIssues)
```

From looking at the pandas stubs (i have pandas-stubs 2.0.1.230501 installed.)
https://github.com/pandas-dev/pandas-stubs/blob/v2.0.1.230501/pandas-stubs/core/series.pyi#L776
```
    def apply(
        self,
        func: Callable[..., Scalar | Sequence | set | Mapping],
        convertDType: _bool = ...,
        args: tuple = ...,
        **kwds,
    ) -> Series: ...
    @overload
    def apply(
        self,
        func: Callable[..., Series],
        convertDType: _bool = ...,
        args: tuple = ...,
        **kwds,
    ) -> DataFrame: ...
```

It seems that there are two versions. Once which returns a dataframe where the applied function has to return a series and one that returns a Series where the applied function is less restricted.

It seems that pyright is only looking at the Dataframe version and i am not sure why.

Cheers
Jan-Eric

**Comment by erictraut** [comment]
Could you provide a self-contained, minimal code sample that demonstrates the problem you're seeing? (It's best to assume that we know nothing about pandas or any other specific Python library.)

**Comment by JanEricNitschke** [comment]
To repro:

Create and activate a new venv using Python 3.11
Install the latest pandas (2.0.1) via pip install pandas
Install latest pandas-stubs (pandas-stubs 2.0.1.230501) via pip install pandas-stubs
Create a new project and a new file test.py and paste the below code into it

```
import pandas as pd
dataframe: pd.DataFrame = pd.DataFrame(
    {"col1": [{"subcol1": [1, 2, 3], "subcol2": ["a", "b", "c"]}]}
)
def _dict_to_df(my_dict: dict) -> pd.DataFrame:
    return pd.DataFrame(my_dict)
dataframe["col1"] = dataframe["col1"].apply(_dict_to_df)
```

That gives the following type checking error:

```
 error: Argument of type "(my_dict: dict[Unknown, Unknown]) -> DataFrame" cannot be assigned to parameter "func" of type "(...) -> Series[Unknown]" in function "apply"
    Type "(my_dict: dict[Unknown, Unknown]) -> DataFrame" cannot be assigned to type "(...) -> Series[Unknown]"
      Function return type "DataFrame" is incompatible with type "Series[Unknown]"
        "DataFrame" is incompatible with "Series[Unknown]" (reportGeneralTypeIssues)
```

**Reply by erictraut** [reply]
This appears to be due to the `pandas-stubs` library that you have installed, similar to the [other issue](https://github.com/microsoft/pyright/issues/5136#issuecomment-1554387534) that you reported. So this appears to be a bug in the `pandas-stubs` you're using.

**Reply by JanEricNitschke** [reply]
Thanks, i have created issues on their github page for both situations.
